src/projection.jl

```diff
@@ -135,13 +132,33 @@ ProjectTo(::Real) = ProjectTo{Real}()
 ProjectTo(::Complex) = ProjectTo{Complex}()
 ProjectTo(::Number) = ProjectTo{Number}()
 for T in (Float16, Float32, Float64, ComplexF16, ComplexF32, ComplexF64)
-    # Preserve low-precision floats as accidental promotion is a common perforance bug
+    # Preserve low-precision floats as accidental promotion is a common perforance bug    
     @eval ProjectTo(::$T) = ProjectTo{$T}()
 end
-ProjectTo(x::Integer) = ProjectTo(float(x))
-ProjectTo(x::Complex{<:Integer}) = ProjectTo(float(x))
-(::ProjectTo{T})(dx::Number) where {T<:Number} = convert(T, dx)
-(::ProjectTo{T})(dx::Number) where {T<:Real} = convert(T, real(dx))
+
+# Preserve low-precision floats as accidental promotion is a common perforance bug
+(::ProjectTo{T})(dx::AbstractFloat) where T<:AbstractFloat = convert(T, dx)
+function (proj::ProjectTo{<:Complex{<:AbstractFloat}})(
+    dx::Union{Complex{<:AbstractFloat}, AbstractFloat}
+)
+    # in this case we can just convert as we know we are dealing with 
+    # boring floating point types or integers
+    return convert(project_type(proj), dx)
+end
+
+
+# We asked for a number/real and they gave use one. We did ask for a particular concrete
+# type, but that is just for the preserving low precision floats, which is handled above.
+# Any Number/Real actually occupies the same subspace, so we can trust them.
+# In particular, this makes weirder Real subtypes that are not simply the values like
+# ForwardDiff.Dual and Symbolics.Sym work, because we stay out of their way.
+(::ProjectTo{<:Number})(dx::Number) = dx
+(::ProjectTo{<:Real})(dx::Real) = dx
+
+(::ProjectTo{T})(dx::Complex) where T<:Real = ProjectTo(zero(T))(real(dx))
```

Comment:
No need to re-create the projector:
```suggestion
(project::ProjectTo{<:Real})(dx::Complex) = project(real(dx))
```

Reply:
true, good catch